src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java

```diff
@@ -15,102 +15,568 @@
+   * Calls the specified function with retries and backoffs until the retries are exhausted or the
+   * function succeeds.
    *
-   * @param indices the set of index names to create, not null
+   * @param description description of the attempted action n present tense
+   * @param function the function to call and retry
+   * @param <T> the return type of the function
+   * @return the return value of the called function
    */
-  void createIndices(Set<String> indices);
+  private <T> T callWithRetries(String description, Callable<T> function) {
+    try {
+      return RetryUtil.callWithRetries(
+          description,
+          function,
+          config.maxRetries(),
+          config.retryBackoffMs()
+      );
+    } catch (Exception e) {
+      throw new ConnectException("Failed to " + description + ".", e);
+    }
+  }
 
   /**
-   * Creates an explicit mapping.
+   * Configures HTTP authentication and proxy authentication according to the client configuration.
    *
-   * @param index the index to write
-   * @param type the type for which to create the mapping
-   * @param schema the schema used to infer the mapping
-   * @throws IOException if the client cannot execute the request
+   * @param builder the HttpAsyncClientBuilder
    */
-  void createMapping(String index, String type, Schema schema) throws IOException;
+  private void configureAuthentication(HttpAsyncClientBuilder builder) {
+
+    CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+    if (config.isAuthenticatedConnection()) {
+      config.connectionUrls().forEach(
+          url ->
+              credentialsProvider.setCredentials(
+                  new AuthScope(new HttpHost(url)),
+                  new UsernamePasswordCredentials(config.username(), config.password().value())
+              )
+      );
+      builder.setDefaultCredentialsProvider(credentialsProvider);
+    }
+
+    if (config.isBasicProxyConfigured()) {
+      HttpHost proxy = new HttpHost(config.proxyHost(), config.proxyPort());
+      builder.setProxy(proxy);
+
+      if (config.isProxyWithAuthenticationConfigured()) {
+        credentialsProvider.setCredentials(
+            new AuthScope(proxy),
+            new UsernamePasswordCredentials(config.proxyUsername(), config.proxyPassword().value())
+        );
+      }
+
+      builder.setDefaultCredentialsProvider(credentialsProvider);
+    }
+  }
 
   /**
-   * Gets the JSON mapping for the given index and type. Returns {@code null} if it does not exist.
+   * Configures the client to use SSL if configured.
    *
-   * @param index the index
-   * @param type the type
-   * @throws IOException if the client cannot execute the request
+   * @param builder the HttpAsyncClientBuilder
    */
-  JsonObject getMapping(String index, String type) throws IOException;
+  private void configureSslContext(HttpAsyncClientBuilder builder) {
+    SslFactory kafkaSslFactory = new SslFactory(Mode.CLIENT, null, false);
```

Comment:
We are using this SSL context to talk to ES, not to Kafka, right? The variable name is confusing.

Reply:
yes to ES, but this class is provided by `Kafka` hence the naming. this is ported from the `JestClient`, but can change to `sslFactory` if you feel it is necessary

Reply:
IMO it's less confusing if you call it `sslFactory`.